### PR TITLE
Port the HPMC remove_drift updater to v.3.0 API

### LINKED ISSUE #514: Move HPMC remove_drift updater to v.3.0 API
## Description
Refactor the HPMC remove_drift updater to the HOOMD v.3.0 API. The template can be found in the `feature/new-object-API` branch's `Clusters` updater in `hoomd/hpmc/update.py`.

## Proposed solution
The solution involves following the template mentioned above.

## Additional context
Any pull request should merge into `feature/new-object-API` **not** `master`.

### PR DESCRIPTION
**Pull request by lyrivera**
## Description

This PR ports the HPMC remove_drift updater to v.3.0 API

## Motivation and context

Resolves #514 

## How has this been tested?

Not tested yet.

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/master/ContributorAgreement.md).
- [x] My name is on the list of contributors (`sphinx-doc/credits.rst`) in the pull request source branch.

### CODE REVIEW

File: hoomd/hpmc/UpdaterRemoveDrift.h [hunk 1]
```diff
@@ -43,10 +42,46 @@ template<class Shape> class RemoveDriftUpdater : public Updater
...
+                       std::shared_ptr<IntegratorHPMCMono<Shape>> mc,
+                       pybind11::list ref_positions)
+        : Updater(sysdef), m_mc(mc)
         {
+        m_ref_positions.resize(m_pdata->getN());
+        setRefPositions(ref_positions);
+        }
+
+    //! Get reference positions as a python list of 3-tuples
+    pybind11::list getRefPositions()
```
Comment (b-butler) [inline hoomd/hpmc/UpdaterRemoveDrift.h:54]: Given that this is `N` particles long. I think accepting a NumPy array would be preferable here, as that is typically how we store large numerical data (roughly the size of the simulation) versus smaller amounts of data like a shape definition or pair parameter parametrization.

File: hoomd/hpmc/update.py [hunk 2]
```diff
@@ -490,72 +490,56 @@ def N(self):  # noqa: N802 - allow N as a function name
...
 
     Args:
-        mc (:py:mod:`hoomd.hpmc.integrate`): MC integrator.
-        external_lattice (:py:class:`hoomd.hpmc.field.lattice_field`): lattice
-          field where the lattice is defined.
-        period (int): the period to call the updater
-
-    The command hpmc.update.remove_drift sets up an updater that removes the
-    center of mass drift of a system every period timesteps,
+        ref_positions (`list` [`tuple` [`float`, `float`]], **required**) - the
```
Comment (b-butler) [inline hoomd/hpmc/update.py:497]: This presumes 2D currently since the list says two floats. I would prefer to see this make a NumPy array.

File: hoomd/hpmc/update.py [hunk 3]
```diff
@@ -490,72 +490,56 @@ def N(self):  # noqa: N802 - allow N as a function name
...
-            else:
-                hoomd.context.current.device.cpp_msg.error(
-                    "update.remove_drift: Unsupported integrator.\n")
-                raise RuntimeError("Error initializing update.remove_drift")
-        else:
-            raise RuntimeError(
-                "update.remove_drift: Error! GPU not implemented.")
+    def __init__(self, ref_positions, trigger=1):
+        super().__init__(trigger)
+        _param_dict = dict(ref_positions=list(ref_positions), trigger=trigger)
```
Comment (b-butler) [inline hoomd/hpmc/update.py:513]: `trigger` is not needed in the dict here. Also, no validation of the `ref_positions` is being done here.

File: hoomd/hpmc/update.py [hunk 4]
```diff
@@ -490,72 +490,56 @@ def N(self):  # noqa: N802 - allow N as a function name
...
-                               external_lattice.cpp_compute, mc.cpp_integrator)
-        self.setupUpdater(period)
+    def _add(self, simulation):
+        """Add the operation to a simulation.
+
+        HPMC uses RNGs. Warn the user if they did not set the seed.
+        """
+        if simulation is not None:
+            simulation._warn_if_seed_unset()
+
```
Comment (b-butler) [inline hoomd/hpmc/update.py:523]: This particular operation does not use the seed as far as I am aware though, so this warning is not needed.

File: hoomd/hpmc/update.py [hunk 5]
```diff
@@ -490,72 +490,56 @@ def N(self):  # noqa: N802 - allow N as a function name
...
+        integrator = self._simulation.operations.integrator
+        if not isinstance(integrator, integrate.HPMCIntegrator):
+            raise RuntimeError("The integrator must be a HPMC integrator.")
+
+        if not integrator._attached:
+            raise RuntimeError("Integrator is not attached yet.")
+
+        if isinstance(self._simulation.device, hoomd.device.GPU):
+            raise RuntimeError("GPU not supported.")
+
```
Comment (b-butler) [inline hoomd/hpmc/update.py:536]: We can just update on the CPU with a `UserWarning` about performance right?

File: hoomd/hpmc/pytest/test_remove_drift.py [hunk 6]
```diff
@@ -0,0 +1,140 @@
+# Copyright (c) 2009-2021 The Regents of the University of Michigan
```
Comment (b-butler) [inline hoomd/hpmc/pytest/test_remove_drift.py:1]: Every test is restricted to serial. Does the `UpdaterRemoveDrift` not work in MPI simulations?

**Review by b-butler — CHANGES_REQUESTED** [review]
I have a few comments. Currently, I am not sure why we restrict all our test to serial runs of pytest, and I would like to see ref_positions` be moved to a NumPy array, and be called `reference_positions`. We should probably return the array by copy and document it as such to prevent users from causing memory errors and having unexpected effects.

File: hoomd/hpmc/UpdaterRemoveDrift.h (hunk 1 above)
Comment (lyrivera) [reply hoomd/hpmc/UpdaterRemoveDrift.h:54]: Makes sense but I'm not sure how to handle Numpy arrays with ParameterDict. Is this done in some other place that I can take as an example?

Comment (b-butler) [reply hoomd/hpmc/UpdaterRemoveDrift.h:54]: No, I don't believe we have had an example of this yet, but I can whip up a type validator for `numpy.ndarray` fairly easily, and will ping you when done.

Comment (lyrivera) [reply hoomd/hpmc/UpdaterRemoveDrift.h:54]: Great, thanks.

**Comment by joaander** [comment]
@lyrivera have you made progress on responding to this review?

Also, there have been a number of changes on master recently. I recommend merging master up into your branch regularly to keep up with recent developments.

**Comment by lyrivera** [comment]
> @lyrivera have you made progress on responding to this review?
> 
> Also, there have been a number of changes on master recently. I recommend merging master up into your branch regularly to keep up with recent developments.

Yes. Sorry for the delay, I will have it ready for another review by the end of the day.

**Comment by lyrivera** [comment]
@b-butler, @yuanzhou0827  I believe this is ready for another review. The pickling test is still failing because the element-wise  comparison that the `==` operator performs on Numpy arrays. I believe this can be fixed by overloading the `__eq__` method of the `ParameterDict` class but I'm not sure.

**Review by b-butler — CHANGES_REQUESTED** [review]
I have a few comments to address, and you are right, we need to modify the `ParameterDict.__eq__` to handle array quantities correctly.

File: hoomd/hpmc/update.py [hunk 7]
```diff
@@ -466,72 +467,51 @@ def N(self):  # noqa: N802 - allow N as a function name
...
 
     Args:
-        mc (:py:mod:`hoomd.hpmc.integrate`): MC integrator.
-        external_lattice (:py:class:`hoomd.hpmc.field.lattice_field`): lattice
-          field where the lattice is defined.
-        period (int): the period to call the updater
-
-    The command hpmc.update.remove_drift sets up an updater that removes the
-    center of mass drift of a system every period timesteps,
+        reference_positions ((*N*, 3) `numpy.ndarray` of ``numpy.float32``): the
```
[the remark this answers is not in the document]
Comment (joaander) [reply hoomd/hpmc/update.py:474]: The API should not depend on the build configuration. Don't use `Scalar` for datatypes exposed to the API, use `double`.

File: hoomd/hpmc/update.py [hunk 8]
```diff
@@ -466,72 +467,51 @@ def N(self):  # noqa: N802 - allow N as a function name
...
+        integrator = self._simulation.operations.integrator
+        if not isinstance(integrator, integrate.HPMCIntegrator):
+            raise RuntimeError("The integrator must be a HPMC integrator.")
+
+        if not integrator._attached:
+            raise RuntimeError("Integrator is not attached yet.")
+
+        if isinstance(self._simulation.device, hoomd.device.GPU):
+            self._simulation.device._cpp_msg.warning(
+                "Falling back on CPU. No GPU implementation available.\n")
```
[the remark this answers is not in the document]
Comment (joaander) [reply hoomd/hpmc/update.py:507]: We have one case where we use `warnings.warn`: the deprecation of `Snapshot.exists`. I don't know that we have any concrete plans to replace the Messenger class.

File: hoomd/hpmc/update.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (lyrivera) [reply hoomd/hpmc/update.py:474]: @joaander Do you mean that the `reference_positions` should also be stored internally as `double` instead of `Scalar`?  My confusion comes from the observation that there's other cases of exported properties to the python API that are stored internally as `Scalar`. Or perhaps  I'm misunderstanding?

[the remark this answers is not in the document]
Comment (joaander) [reply hoomd/hpmc/update.py:474]: Does the param dict reference the array by copy or by reference?

[the remark this answers is not in the document]
Comment (lyrivera) [reply hoomd/hpmc/update.py:474]: I'm not entirely sure. @b-butler is better suited to answer this.

[the remark this answers is not in the document]
Comment (joaander) [reply hoomd/hpmc/update.py:474]: OK, I looked at the code. `getReferencePositions` returns a copy. So you can modify `setReferencePositions` and `getReferencePositions` to work with `double` explicitly even while the internal data array is stored as a `Scalar` (floats can be cast to double with no data loss). This will keep the API consistent going forward when we enable mixed precision computations and the internal buffer will also be in double - but you don't need to make that change now.

[the remark this answers is not in the document]
Comment (lyrivera) [reply hoomd/hpmc/update.py:474]: Ah, I see what you mean now. Thanks

[the remark this answers is not in the document]
Comment (joaander) [reply hoomd/hpmc/update.py:474]: Then you'll also need to update the documentation to `float64` as the data type.

File: hoomd/hpmc/update.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (lyrivera) [reply hoomd/hpmc/update.py:509]: Leaving as is for now until a concrete plan is devised .

**Comment by lyrivera** [comment]
@b-butler @yuanzhou0827 @joaander This is ready for a final review. All test should pass now.

File: hoomd/hpmc/update.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (b-butler) [reply hoomd/hpmc/update.py:509]: I was thinking of how we deal with exceptions and error messages. We have moved the error message to the exception in multiple cases. We can leave this here though if that is what is decided. It is just not catch-able by users in Python then.

**Comment by lyrivera** [comment]
@b-butler There is still one last failing test that I haven't been able to figure out. Any help would much appreciated.

**Comment by b-butler** [comment]
@lyrivera could you add a test that tests the behavior of `UpdaterRemoveDrift::update`?

**Comment by lyrivera** [comment]
> @lyrivera could you add a test that tests the behavior of `UpdaterRemoveDrift::update`?

Sure.

File: hoomd/hpmc/UpdaterRemoveDrift.h [hunk 9]
```diff
@@ -43,10 +42,49 @@ template<class Shape> class RemoveDriftUpdater : public Updater
...
         {
+        setReferencePositions(ref_positions);
+        }
+
+    //! Set reference positions from a Nx3 numpy array
+    void setReferencePositions(const pybind11::array_t<double> ref_pos)
+        {
+        const size_t N = ref_pos.request().shape[0];
+        const size_t dim = ref_pos.request().shape[1];
+        if (N != this->m_pdata->getN() || dim != 3)
```
Comment (lyrivera) [inline hoomd/hpmc/UpdaterRemoveDrift.h:57]: @b-butler I believe we need `getNGlobal()` here or am I missing something?

File: hoomd/hpmc/UpdaterRemoveDrift.h [hunk 10]
```diff
@@ -43,10 +42,49 @@ template<class Shape> class RemoveDriftUpdater : public Updater
...
+            {
+            throw std::runtime_error("The array must be of shape (N_particles, 3).");
+            }
+        const double* rawdata = static_cast<double*>(ref_pos.request().ptr);
+        m_ref_positions.resize(m_pdata->getN());
```
Comment (lyrivera) [inline hoomd/hpmc/UpdaterRemoveDrift.h:62]: @b-butler same here

File: hoomd/hpmc/UpdaterRemoveDrift.h (hunk 9 above)
Comment (b-butler) [reply hoomd/hpmc/UpdaterRemoveDrift.h:57]: Likely so, unless we want to deal with particle migration.

File: hoomd/hpmc/UpdaterRemoveDrift.h (hunk 10 above)
Comment (b-butler) [reply hoomd/hpmc/UpdaterRemoveDrift.h:62]: Yes

File: hoomd/hpmc/update.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (joaander) [reply hoomd/hpmc/update.py:511]: I was trying to say that we haven't decided anything yet, so this is fine as it is in this PR. We can bring up the topic of how to better offer warnings to user in a future conversation and update them in bulk in another PR.

File: hoomd/data/parameterdicts.py [hunk 11]
```diff
@@ -345,6 +346,13 @@ def __len__(self):
         """int: The number of keys."""
         return len(self._dict)
 
+    def __eq__(self, other):
+        """Equality between ParameterDict objects."""
+        if not isinstance(other, ParameterDict):
+            return NotImplemented
+        return (set(self.keys()) == set(other.keys()) and np.all(
+            np.all(self[key] == other[key]) for key in self.keys()))
```
Comment (joaander) [inline hoomd/data/parameterdicts.py:354]: I have the same question on this PR as in #965. Is it valid to use numpy for this check on keys that may not be numpy arrays?

File: hoomd/hpmc/UpdaterRemoveDrift.h [hunk 12]
```diff
@@ -43,10 +42,49 @@ template<class Shape> class RemoveDriftUpdater : public Updater
...
+        {
+        std::vector<size_t> dims(2);
+        dims[0] = this->m_ref_positions.size();
+        dims[1] = 3;
+        // the cast from vec3<Scalar>* to Scalar* is safe since vec3 is tightly packed without any
+        // padding. This also makes a copy so, modifications of this array do not effect the
+        // original reference positions.
+        const auto reference_array = pybind11::array_t<double>(
+            dims,
+            reinterpret_cast<const double*>(&(this->m_ref_positions[0])));
```
Comment (joaander) [inline hoomd/hpmc/UpdaterRemoveDrift.h:83]:
```suggestion
            reinterpret_cast<const double*>(this->m_ref_positions.data()));
```
Cleaner access to the data pointer.

File: hoomd/hpmc/UpdaterRemoveDrift.h [hunk 13]
```diff
@@ -43,10 +42,49 @@ template<class Shape> class RemoveDriftUpdater : public Updater
...
+        const double* rawdata = static_cast<double*>(ref_pos.request().ptr);
+        m_ref_positions.resize(m_pdata->getNGlobal());
+        for (size_t i = 0; i < N; i++)
+            {
+            const size_t array_index = i * 3;
+            this->m_ref_positions[i] = vec3<Scalar>(rawdata[array_index],
+                                                    rawdata[array_index + 1],
+                                                    rawdata[array_index + 2]);
+            }
+        }
```
Comment (joaander) [inline hoomd/hpmc/UpdaterRemoveDrift.h:70]: In MPI builds, broadcast m_ref_positions from the root rank to all ranks and ignore values set on non-root ranks.

File: hoomd/hpmc/pytest/test_remove_drift.py [hunk 14]
```diff
@@ -0,0 +1,189 @@
...
+        trigger=hoomd.trigger.Periodic(1),
+        reference_positions=reference_positions)
+    sim.operations.updaters.append(remove_drift)
+    sim.run(1)
+
+    s = sim.state.snapshot
+    if s.communicator.rank == 0:
+        reference_com = np.mean(reference_positions, axis=0)
+        new_com = np.mean(s.particles.position, axis=0)
+        assert np.allclose(reference_com, new_com, atol=0.05)
```
Comment (joaander) [inline hoomd/hpmc/pytest/test_remove_drift.py:174]: This test is failing. Given the periodic boundary conditions, I don't think that the remove drift updater leads to a conservation of the center of mass of the system.

To test, you need to compute the drift (you can avoid periodic boundary conditions by unwrapping positions) from the initial lattice position and ensure that the drift is close to 0.

File: hoomd/hpmc/pytest/test_remove_drift.py [hunk 15]
```diff
@@ -0,0 +1,189 @@
...
+    sim.operations.integrator = mc
+
+    # use initial lattice configuration as reference
+    comm = MPI.COMM_WORLD
+    s = sim.state.snapshot
+    if s.communicator.rank == 0:
+        reference_positions = s.particles.position
+    else:
+        reference_positions = None
+    reference_positions = comm.bcast(reference_positions, root=0)
```
Comment (joaander) [inline hoomd/hpmc/pytest/test_remove_drift.py:158]: Implement the broadcast in C++. Users should not have to use mpi4py to ensure that they specify the same lattice positions on all ranks.

File: hoomd/hpmc/update.py [hunk 16]
```diff
@@ -466,72 +467,54 @@ def N(self):  # noqa: N802 - allow N as a function name
...
     Args:
-        mc (:py:mod:`hoomd.hpmc.integrate`): MC integrator.
-        external_lattice (:py:class:`hoomd.hpmc.field.lattice_field`): lattice
-          field where the lattice is defined.
-        period (int): the period to call the updater
-
-    The command hpmc.update.remove_drift sets up an updater that removes the
-    center of mass drift of a system every period timesteps,
+        reference_positions ((*N*, 3) `numpy.ndarray` of ``numpy.float64``): the
+            reference positions of the lattice.
```
Comment (joaander) [inline hoomd/hpmc/update.py:475]:
```suggestion
            reference positions of the lattice :math:`[\\mathrm{length}]`.
```

**Review by joaander — CHANGES_REQUESTED** [review]
Thanks, this needs a few more changes before it is ready to merge.

File: hoomd/data/parameterdicts.py (hunk 11 above)
Comment (b-butler) [reply hoomd/data/parameterdicts.py:354]: Yes, this should be valid. When passed a Boolean returns that Boolean value, and when passed a sequence computes the n-wise conjunction as expected.

File: hoomd/hpmc/pytest/test_remove_drift.py (hunk 15 above)
Comment (lyrivera) [reply hoomd/hpmc/pytest/test_remove_drift.py:158]: This was just a workaround to be used in this test only since I'm not setting the reference positions explicitly on python. Instead, I'm using whatever initial positions are set by the `lattice_snapshot_factory`, which I can only get through the snapshot API. I can avoid it by setting the positions explicitly, I believe.

File: hoomd/hpmc/pytest/test_remove_drift.py (hunk 14 above)
Comment (lyrivera) [reply hoomd/hpmc/pytest/test_remove_drift.py:174]: I see. I thought it was related to the particle tags.

File: hoomd/hpmc/pytest/test_remove_drift.py (hunk 15 above)
Comment (joaander) [reply hoomd/hpmc/pytest/test_remove_drift.py:158]: I still think the array should be broadcast in c++ to ensure consistency. Using the snapshot to get the initial lattice is likely a common use-case. Even if users don't use the snapshot, they should not be forced to duplicate the data on all ranks to ensure the simulation is valid.

File: hoomd/hpmc/pytest/test_remove_drift.py (hunk 14 above)
Comment (lyrivera) [reply hoomd/hpmc/pytest/test_remove_drift.py]: @joaander I changed the test to compute the drift from the original configuration (after unwrapping) and check if its close to zero but it keeps failing. I've checked the source code for any obvious issues but nothing stands out to me.

Comment (joaander) [reply hoomd/hpmc/pytest/test_remove_drift.py]: To understand why this test is failing, I would suggest inspecting the before and after remove drift configurations in a visualization tool.

File: hoomd/hpmc/pytest/test_remove_drift.py (hunk 15 above)
Comment (joaander) [reply hoomd/hpmc/pytest/test_remove_drift.py:158]: Now that you've added the broadcast in c++, you can remove the broadcast in the python test.

File: hoomd/hpmc/pytest/test_remove_drift.py (hunk 6 above)
Comment (joaander) [reply hoomd/hpmc/pytest/test_remove_drift.py:1]: This is not resolved, one test still is marked `@skip_mpi`.

**Review by joaander — COMMENTED** [review]
This is not quite ready to merge. There are several unresolved conversations and a unit test is failing.

File: hoomd/hpmc/update.py [hunk 17]
```diff
@@ -466,72 +467,54 @@ def N(self):  # noqa: N802 - allow N as a function name
...
-
-    The command hpmc.update.remove_drift sets up an updater that removes the
-    center of mass drift of a system every period timesteps,
+        reference_positions ((*N*, 3) `numpy.ndarray` of ``numpy.float64``): the
+            reference positions of the lattice :math:`[\\mathrm{length}]`.
+        trigger (`hoomd.trigger.Trigger`): Select the timesteps to remove drift.
 
-    Example::
+    During the time steps specified by *trigger*, particle positions are
+    modified such that the their center of mass coincides with that of
```
Comment (joaander) [inline hoomd/hpmc/update.py:479]: Correct the phrasing here. The center of mass does not enter into this operation.

Additionally, following the guidance of #1078, please add LaTeX math that describes the computation of the drift. Use the same notation as #1079 for consistency (here, `\vec{r}_i` for particle positions and `\mathrm{N_particles}` for the number of particles). I haven't introduced notation for the box minimum image yet, one idea would be `\mathrm{min_image}(\vec{a})`.

**Review by joaander — CHANGES_REQUESTED** [review]
@lyrivera what change led to the test passing? I don't see anything obvious.

Regardless, I made many suggestions to simplify the test, remove unnecessary code, and improve the documentation.

File: hoomd/hpmc/UpdaterRemoveDrift.h [hunk 18]
```diff
@@ -113,8 +155,8 @@ template<class Shape> class RemoveDriftUpdater : public Updater
         }
 
     protected:
-    std::shared_ptr<ExternalFieldLattice<Shape>> m_externalLattice;
     std::shared_ptr<IntegratorHPMCMono<Shape>> m_mc;
+    std::vector<vec3<Scalar>> m_ref_positions;
```
[the remark this answers is not in the document]
Comment (lyrivera) [reply hoomd/hpmc/UpdaterRemoveDrift.h:159]: I agree. In that case, should we consider removing the Shape template altogether and, perhaps, moving the updater to a different namespace, like `hoomd.update.RemoveDrift`?

[the remark this answers is not in the document]
Comment (joaander) [reply hoomd/hpmc/UpdaterRemoveDrift.h:159]: @b-butler what do you think?

[the remark this answers is not in the document]
Comment (b-butler) [reply hoomd/hpmc/UpdaterRemoveDrift.h:159]: I would say if there is any utility to the updater in MD then use the more general namespace (I am not sure there is), and if not leave it in HPMC. If we were to add a new MC move type that did not directly rely on the integrator, we would still put that under `hoomd.hpmc` (I think), so I feel like other updaters should be the same.

However, if MD could use the updater productively in some cases, then the more general namespace should be used for visibility since many people who use `hoomd.md` may never use `hoomd.hpmc`.

[the remark this answers is not in the document]
Comment (joaander) [reply hoomd/hpmc/UpdaterRemoveDrift.h:159]: This method is to improve the accuracy of FrenkelLladd calculations. Those can be performed in MD as well as MC, so there is at least one case where a user may want to use it outside of HPMC.

[the remark this answers is not in the document]
Comment (lyrivera) [reply hoomd/hpmc/UpdaterRemoveDrift.h:159]: Since it doesn't require integrator information and there is at least one use case, I believe it should be placed a more general namespace. However, I'll wait for consensus before proceeding. On another note: I have already removed the shape template and have addressed a bunch of the open comments locally.

[the remark this answers is not in the document]
Comment (b-butler) [reply hoomd/hpmc/UpdaterRemoveDrift.h:159]: Okay, then we should move to the more general namespace.

[the remark this answers is not in the document]
Comment (lyrivera) [reply hoomd/hpmc/UpdaterRemoveDrift.h:153]: Sounds good. I'll work on it.

**Comment by lyrivera** [comment]
@joaander @b-butler I believe this is ready for another pass.

File: hoomd/update/remove_drift.py [hunk 19]
```diff
@@ -0,0 +1,52 @@
...
+from hoomd import _hoomd
+import numpy as np
+
+
+class RemoveDrift(Updater):
+    r"""Remove the average drift from a system restrained on a lattice.
+
+    Args:
+        reference_positions ((*N_particles*, 3) `numpy.ndarray` of
+            ``numpy.float64``): the reference positions of the
```
Comment (joaander) [inline hoomd/update/remove_drift.py:19]:
```suggestion
            `float`): the reference positions of the
```
@b-butler is recommending using `float` for numpy array data types that store double precision floating point values.

File: sphinx-doc/module-hoomd-update.rst [hunk 20]
```diff
@@ -9,11 +9,12 @@ hoomd.update
...
     BoxResize
+    RemoveDrift
     CustomUpdater
 
 .. rubric:: Details
 
 .. automodule:: hoomd.update
     :synopsis: Modify the system state periodically.
-    :members: BoxResize, CustomUpdater
+    :members: BoxResize, RemoveDrift, CustomUpdater
```
Comment (joaander) [inline sphinx-doc/module-hoomd-update.rst:19]:
```suggestion
    :members: BoxResize, CustomUpdater, RemoveDrift
```
Alphabetize.

File: sphinx-doc/module-hoomd-update.rst [hunk 21]
```diff
@@ -9,11 +9,12 @@ hoomd.update
     :nosignatures:
 
     BoxResize
+    RemoveDrift
     CustomUpdater
```
Comment (joaander) [inline sphinx-doc/module-hoomd-update.rst:13]:
```suggestion
    CustomUpdater
    RemoveDrift
```
Alphabetize.

**Review by joaander — CHANGES_REQUESTED** [review]
Thanks! This is looking a lot better now. I made a few suggestions to make the documentation consistent with the rest of HOOMD-blue. 

@yuanzhou0827, @b-butler, do you have any other suggestions?

**Review by b-butler — CHANGES_REQUESTED** [review]
I have one comment about the copying of reference positions to expose to Python that could cause errors in single precision mode.

File: hoomd/UpdaterRemoveDrift.h [hunk 22]
```diff
@@ -0,0 +1,153 @@
...
+        {
+        std::vector<size_t> dims(2);
+        dims[0] = this->m_ref_positions.size();
+        dims[1] = 3;
+        // the cast from vec3<Scalar>* to Scalar* is safe since vec3 is tightly packed without any
+        // padding. This also makes a copy so, modifications of this array do not effect the
+        // original reference positions.
+        const auto reference_array = pybind11::array_t<double>(
+            dims,
+            reinterpret_cast<const double*>(this->m_ref_positions.data()));
```
[the remark this answers is not in the document]
Comment (joaander) [reply hoomd/UpdaterRemoveDrift.h:74]: Good catch. Store then as `Scalar` and don't cast or copy on return. The data types in the documentation naming double will be consistent with the default and supported build types.

Later, when we implement mixed precision, `Scalar` will be changed to the type that matches the particle positions.

**Comment by joaander** [comment]
@lyrivera Please merge `master` into this branch and address the change that @b-butler requested. Then this is ready for a final review.

File: hoomd/UpdaterRemoveDrift.h (hunk 22 above)
[the remark this answers is not in the document]
Comment (lyrivera) [reply hoomd/UpdaterRemoveDrift.h]: @joaander I tried removing the `reinterpret_cast` (f that what you meant) but I was getting compilation errors due to a mismatch of types in the `array_t` constructor. I couldn't figure it out (tbh didn't spend too much time in it) so I'm simply casting const Scalar * as @b-butler  suggested.

[the remark this answers is not in the document]
Comment (joaander) [reply hoomd/UpdaterRemoveDrift.h]: This is expected. `m_ref_positions` is an array of `vec3<Scalar>`. It must be cast to be used in `array_t`.

**Review by joaander — APPROVED** [review]

**Review by b-butler — APPROVED** [review]
LGTM! Thanks @lyrivera!
